File: Sources/SkiaKit/BackendSkia/ScContext.swift
import SkiaKitCxx

@_expose(Cxx)
public struct _ScAffineTransform{
    var ctm: CGAffineTransform = .init()
    public init(){}
}

extension ScPaint{
    public mutating func setColor(_ color: CGColor){
      let srgb = color.toSRGBColor()
      self.setColor(srgb.red, srgb.green, srgb.blue, srgb.opacity)
    }
}
extension CGContext{
    public func strokePath<T: _PathConvertible>(convertible path: T, style: StrokeStyle = .init()){
        var paint = ScPaint.init()
        paint.setColor(_strokeColor.withAlphaComponent(_alpha))
        paint.setStroke(true)
        paint.setAntiAlias(true)

        var scpath = ScPath.init()
        scpath.addPath(path)
        _scContext.drawPath(scpath, paint)
    }
    public func fillPath<T: _PathConvertible>(convertible path: T, style: FillStyle = .init()){
        var paint = ScPaint.init()
        paint.setColor(_fillColor.withAlphaComponent(_alpha))
        paint.setStroke(false)
        paint.setAntiAlias(true)

        var scpath = ScPath.init()
        scpath.addPath(path)
        _scContext.drawPath(scpath, paint)
    }
    public func clipPath<T: _PathConvertible>(convertible path: T){
        var scpath = ScPath.init()
        scpath.addPath(path)
        _scContext.clipPath(scpath, true)
    }
    func drawGlyphs(_ count: Int, glyphs: Array<SkGlyphID>, positions: Array<ScPoint>, origin: ScPoint, font: NSFont){
        var paint=ScPaint.init()
        paint.setColor(_fillColor.withAlphaComponent(_alpha))
        paint.setStroke(false)
        paint.setAntiAlias(true)
        _scContext.drawGlyphs(Int32(count), glyphs, positions, origin, font._internal, paint)
    }
    func drawImage(_ image: NSImage, in rect: CGRect, blendMode: CGBlendMode, alpha: CGFloat){
        guard let rep = image.representations.first else{
            return
        }
        if let scimage = rep as? ScImage{
            _scContext.drawImage(scimage, rect.toSkRect());
        }
    }
}
extension ScPath{
    mutating func addPath<P: _PathConvertible>(_ path: P){
        path.forEach { e in
            switch e {
            case let .move(to: p): self.moveTo(SkScalar(p.x), SkScalar(p.y))
            case let .line(to: p): self.lineTo(SkScalar(p.x), SkScalar(p.y))
            case let .quadCurve(to: to, control: cp):
                self.quadTo(cp.toSkPoint(), to.toSkPoint())
            case let .curve(to: to, control1: cp1, control2: cp2):
                self.cubicTo(cp1.toSkPoint(), cp2.toSkPoint(), to.toSkPoint())
                break
            case .closeSubpath: self.close()
            }
        }
    }
}
public protocol _PathConvertible{
    func forEach(_ body: (CGPath.Element) -> Void)
}

extension CGPath: _PathConvertible{
}

typealias SkUnichar = SkiaKitCxx.SkUnichar
typealias SkScalar = SkiaKitCxx.SkScalar
typealias ScFloat = SkiaKitCxx.ScFloat
typealias SkGlyphID = SkiaKitCxx.SkGlyphID
typealias ScString = SkiaKitCxx.ScString
typealias ScPoint = SkiaKitCxx.ScPoint
typealias ScSize = SkiaKitCxx.ScSize
typealias ScRect = SkiaKitCxx.ScRect
typealias SC = SkiaKitCxx.SC

typealias ScContext = SkiaKitCxx.ScContext
typealias ScFont = SkiaKitCxx.ScFont
typealias ScWindow = SkiaKitCxx.ScWindow
typealias ScPaint = SkiaKitCxx.ScPaint
typealias ScData = SkiaKitCxx.ScData
typealias ScBitmap = SkiaKitCxx.ScBitmap
typealias ScImage = SkiaKitCxx.ScImage
typealias ScImageInfo = SkiaKitCxx.ScImageInfo
typealias ScColorType = SkiaKitCxx.ScColorType
typealias ScAlphaType = SkiaKitCxx.ScAlphaType
typealias ScFontStyle = SkiaKitCxx.ScFontStyle

File: Sources/SkiaKitCxx/ScContext.cpp
#include "ScContext.h"
#include "SkiaCxxPrivate.h"

ScCanvas::ScCanvas(SkCanvas *c) {
    canvas = c;
}
void ScCanvas::save() const {
    canvas->save();
}
void ScCanvas::restore() const {
    canvas->restore();
}
void ScCanvas::draw(const ScPath &path, const ScPaint &paint) const {
    canvas->drawPath(path.ref(), paint.ref());
}
// void ScCanvas::drawGlyphs(int count, const SkGlyphID glyphs[], const ScPoint positions[], const ScPoint &origin, const ScFont& font, const ScPaint& paint) const{
//     canvas->drawGlyphs(count, glyphs, positions, origin, font.ref(), paint.ref());
// }
// // void ScCanvas::draw(const SkString& text, const ScFont& font, const SkPaint& paint) const{
// //     auto f=font.getFont();
// //     if (f == nullptr){
// //         return;
// //     }
// //     canvas->drawSimpleText(text.data(), text.size(),SkTextEncoding::kUTF8,0,0,*f,paint);
// // }
// void ScCanvas::drawImage(const ScBitmap &bitmap, const SkRect &rect) const{
//     drawImage(bitmap.asImage(), rect);
// }
// void ScCanvas::drawImage(const ScImage &image, const ScRect &rect) const{
//     drawImage(image.getImage(), rect);
// }
// void ScCanvas::concat(SkMatrix &matrix) const {
//     canvas->concat(matrix);
// }
// void ScCanvas::clipPath(const ScPath& path, bool antialiased) const{
//     canvas->clipPath(path.ref(), SkClipOp::kIntersect, antialiased);
// }
void ScPath::moveTo(SkScalar x, SkScalar y){
    ref().moveTo(x, y);
}
void ScPath::lineTo(SkScalar x, SkScalar y){
    ref().lineTo(x, y);
}
void ScPath::quadTo(const ScPoint& p1, const ScPoint& p2){
    ref().quadTo(p1.x(), p1.y(), p2.x(), p2.y());
}
void ScPath::cubicTo(const ScPoint& p1, const ScPoint& p2, const ScPoint& p3){
    ref().cubicTo(p1.x(), p1.y(), p2.x(), p2.y(), p3.x(), p3.y());
}
void ScPath::close(){
    ref().close();
}

void ScPaint::setColor(ScFloat r, ScFloat g, ScFloat b, ScFloat a){
    ref().setColor(SkColor4f{float(r), float(g), float(b), float(a)});
}
void ScPaint::setStroke(bool b){
    ref().setStroke(b);
}
void ScPaint::setAntiAlias(bool b){
    ref().setAntiAlias(b);
}
class ScContext::ScContextInternal{

};
ScContext::ScContext(){

}
ScContext::~ScContext(){
    canvas = nullptr;
}
ScContext::ScContext(const ScContext& c){
    canvas = c.canvas;
}
ScContext::ScContext(SkCanvas *c){
    canvas = c;
}

void ScContext::concatTransform(const float m[6]){
    SkMatrix matrix;
    matrix.setAffine(m);
    canvas->concat(matrix);
}
void ScContext::setTransform(const float m[6]){
    canvas->resetMatrix();
    concatTransform(m);
}
void ScContext::clipPath(const ScPath& path, bool antialiased) const{
    canvas->clipPath(path.ref(), SkClipOp::kIntersect, antialiased);
}

void ScContext::drawPath(const ScPath &path, const ScPaint &paint) const{
    canvas->drawPath(path.ref(), paint.ref());
}
void ScContext::drawGlyphs(int count, const SkGlyphID glyphs[], const ScPoint positions[], const ScPoint &origin, const ScFont& font, const ScPaint& paint) const{
    canvas->drawGlyphs(count, glyphs, (SkPoint*)positions, origin|ScSK, font.ref(), paint.ref());
}
void ScContext::drawImage(const ScImage &image, const ScRect &rect) const{
    canvas->drawImageRect(image.ref(), rect|ScSK, SkSamplingOptions(), nullptr);
}
void ScContext::drawImage(const ScBitmap &bitmap, const ScRect &rect) const{
    canvas->drawImageRect(bitmap.ref().asImage(), rect|ScSK, SkSamplingOptions(), nullptr);
}

void ScContext::save(){
    canvas->save();
}
void ScContext::restore(){
    canvas->restore();
}